File: jambeth/jambeth-cache-bytecode/src/main/java/de/osthus/ambeth/bytecode/behavior/LazyRelationsBehavior.java
package de.osthus.ambeth.bytecode.behavior;

import java.util.List;

import de.osthus.ambeth.bytecode.EmbeddedEnhancementHint;
import de.osthus.ambeth.bytecode.MethodInstance;
import de.osthus.ambeth.bytecode.util.EntityUtil;
import de.osthus.ambeth.bytecode.visitor.EntityMetaDataHolderVisitor;
import de.osthus.ambeth.bytecode.visitor.InterfaceAdder;
import de.osthus.ambeth.bytecode.visitor.RelationsGetterVisitor;
import de.osthus.ambeth.bytecode.visitor.SetCacheModificationMethodCreator;
import de.osthus.ambeth.cache.ValueHolderIEC;
import de.osthus.ambeth.ioc.annotation.Autowired;
import de.osthus.ambeth.log.ILogger;
import de.osthus.ambeth.log.LogInstance;
import de.osthus.ambeth.merge.IEntityMetaDataProvider;
import de.osthus.ambeth.merge.model.IEntityMetaData;
import de.osthus.ambeth.metadata.IEmbeddedMember;
import de.osthus.ambeth.metadata.Member;
import de.osthus.ambeth.metadata.RelationMember;
import de.osthus.ambeth.proxy.IObjRefContainer;
import de.osthus.ambeth.proxy.IValueHolderContainer;
import de.osthus.ambeth.repackaged.org.objectweb.asm.ClassVisitor;
import de.osthus.ambeth.typeinfo.IPropertyInfoProvider;
import de.osthus.ambeth.typeinfo.MethodPropertyInfo;

public class LazyRelationsBehavior extends AbstractBehavior
{
	@SuppressWarnings("unused")
	@LogInstance
	private ILogger log;

	@Autowired
	protected IEntityMetaDataProvider entityMetaDataProvider;

	@Autowired
	protected IPropertyInfoProvider propertyInfoProvider;

	@Autowired
	protected ValueHolderIEC valueHolderContainerHelper;

	@Override
	public Class<?>[] getEnhancements()
	{
		return new Class<?>[] { IObjRefContainer.class, IValueHolderContainer.class };
	}

	@Override
	public ClassVisitor extend(ClassVisitor visitor, IBytecodeBehaviorState state, List<IBytecodeBehavior> remainingPendingBehaviors,
			List<IBytecodeBehavior> cascadePendingBehaviors)
	{
		Class<?> entityType = EntityUtil.getEntityType(state.getContext());
		if (entityType == null)
		{
			return visitor;
		}
		final IEntityMetaData metaData = entityMetaDataProvider.getMetaData(entityType, true);
		if (metaData == null)
		{
			return visitor;
		}
		final boolean addValueHolderContainer;
		if (EmbeddedEnhancementHint.hasMemberPath(state.getContext()))
		{
			for (RelationMember member : metaData.getRelationMembers())
			{
				if (!(member instanceof IEmbeddedMember))
				{
					continue;
				}
				Member cMember = ((IEmbeddedMember) member).getChildMember();
				MethodPropertyInfo prop = (MethodPropertyInfo) propertyInfoProvider.getProperty(cMember.getDeclaringType(), cMember.getName());
				if (state.hasMethod(new MethodInstance(prop.getGetter())) || state.hasMethod(new MethodInstance(prop.getSetter())))
				{
					// Handle this behavior in the next iteration
					cascadePendingBehaviors.add(this);
					return visitor;
				}
			}
			addValueHolderContainer = false;
		}
		else
		{
			for (RelationMember member : metaData.getRelationMembers())
			{
				if (member instanceof IEmbeddedMember)
				{
					continue;
				}
				MethodPropertyInfo prop = (MethodPropertyInfo) propertyInfoProvider.getProperty(member.getDeclaringType(), member.getName());
				if ((prop.getGetter() != null && state.hasMethod(new MethodInstance(prop.getGetter())))
						|| (prop.getSetter() != null && state.hasMethod(new MethodInstance(prop.getSetter()))))
				{
					// Handle this behavior in the next iteration
					cascadePendingBehaviors.add(this);
					return visitor;
				}
			}
			// Add this interface only for real entities, not for embedded types
			addValueHolderContainer = true;
			visitor = new EntityMetaDataHolderVisitor(visitor, metaData);
		}
		visitor = new SetCacheModificationMethodCreator(visitor);
		cascadePendingBehaviors.add(WaitForApplyBehavior.create(beanContext, new WaitForApplyBehaviorDelegate()
		{
			@Override
			public ClassVisitor extend(ClassVisitor visitor, IBytecodeBehaviorState state, List<IBytecodeBehavior> remainingPendingBehaviors,
					List<IBytecodeBehavior> cascadePendingBehaviors)
			{
				if (addValueHolderContainer)
				{
					visitor = new InterfaceAdder(visitor, IValueHolderContainer.class);
				}
				visitor = new RelationsGetterVisitor(visitor, metaData, valueHolderContainerHelper, propertyInfoProvider);
				return visitor;
			}
		}));
		return visitor;
	}
}

File: ambeth/Ambeth.Cache.Bytecode/ambeth/bytecode/behavior/LazyRelationsBehavior.cs
using De.Osthus.Ambeth.Bytecode.Util;
using De.Osthus.Ambeth.Bytecode.Visitor;
using De.Osthus.Ambeth.Cache;
using De.Osthus.Ambeth.Debug;
using De.Osthus.Ambeth.Ioc.Annotation;
using De.Osthus.Ambeth.Log;
using De.Osthus.Ambeth.Merge;
using De.Osthus.Ambeth.Merge.Model;
using De.Osthus.Ambeth.Metadata;
using De.Osthus.Ambeth.Proxy;
using De.Osthus.Ambeth.Typeinfo;
using De.Osthus.Ambeth.Util;
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Reflection;
using System.Reflection.Emit;

namespace De.Osthus.Ambeth.Bytecode.Behavior
{
    public class LazyRelationsBehavior : AbstractBehavior
    {
        public class CascadeBehavior : AbstractBehavior
        {
            protected IEntityMetaData metaData;

            protected bool implementValueHolderContainerInterface;

            [Autowired]
            public IPropertyInfoProvider PropertyInfoProvider { protected get; set; }

            [Autowired]
            public ValueHolderIEC ValueHolderContainerHelper { protected get; set; }

            public CascadeBehavior(IEntityMetaData metaData, bool implementValueHolderContainerInterface)
            {
                this.metaData = metaData;
                this.implementValueHolderContainerInterface = implementValueHolderContainerInterface;
            }

            public override IClassVisitor Extend(IClassVisitor visitor, IBytecodeBehaviorState state, IList<IBytecodeBehavior> remainingPendingBehaviors,
                    IList<IBytecodeBehavior> cascadePendingBehaviors)
            {
                ListUtil.AddAll(cascadePendingBehaviors, 0, remainingPendingBehaviors);
                remainingPendingBehaviors.Clear();

                // Add this interface only for real entities, not for embedded types
                if (implementValueHolderContainerInterface)
                {
                    visitor = new InterfaceAdder(visitor, typeof(IValueHolderContainer));
                }
                visitor = new RelationsGetterVisitor(visitor, metaData, ValueHolderContainerHelper, PropertyInfoProvider);
                visitor = new SetCacheModificationMethodCreator(visitor);
                return visitor;
            }
        }

        [LogInstance]
        public ILogger Log { private get; set; }

        [Autowired]
        public IEntityMetaDataProvider EntityMetaDataProvider { protected get; set; }

        [Autowired]
        public IPropertyInfoProvider PropertyInfoProvider { protected get; set; }

        [Autowired]
        public ValueHolderIEC ValueHolderContainerHelper { protected get; set; }

        public override Type[] GetEnhancements()
        {
            return new Type[] { typeof(IValueHolderContainer) };
        }

        public override IClassVisitor Extend(IClassVisitor visitor, IBytecodeBehaviorState state, IList<IBytecodeBehavior> remainingPendingBehaviors,
                        IList<IBytecodeBehavior> cascadePendingBehaviors)
        {
            Type entityType = EntityUtil.GetEntityType(state.Context);
            if (entityType == null)
            {
                return visitor;
            }
            IEntityMetaData metaData = EntityMetaDataProvider.GetMetaData(entityType, true);
            if (metaData == null)
            {
                return visitor;
            }
            visitor = new FlattenDebugHierarchyVisitor(visitor, metaData.RelationMembers.Length != 0);
            bool addValueHolderContainer;
            if (EmbeddedEnhancementHint.HasMemberPath(state.Context))
            {
                foreach (RelationMember member in metaData.RelationMembers)
                {
                    if (!(member is IEmbeddedMember))
                    {
                        continue;
                    }
                    Member cMember = ((IEmbeddedMember)member).ChildMember;
                    MethodPropertyInfo prop = (MethodPropertyInfo)PropertyInfoProvider.GetProperty(cMember.DeclaringType, cMember.Name);
                    if ((prop.Getter != null && state.HasMethod(new MethodInstance(prop.Getter))) || (prop.Setter != null && state.HasMethod(new MethodInstance(prop.Setter))))
                    {
                        // Handle this behavior in the next iteration
                        cascadePendingBehaviors.Add(this);
                        return visitor;
                    }
                }
                addValueHolderContainer = false;
            }
            else
            {
                foreach (RelationMember member in metaData.RelationMembers)
                {
                    if (member is IEmbeddedMember)
                    {
                        continue;
                    }
                    MethodPropertyInfo prop = (MethodPropertyInfo)PropertyInfoProvider.GetProperty(member.DeclaringType, member.Name);
                    if ((prop.Getter != null && state.HasMethod(new MethodInstance(prop.Getter))) || (prop.Setter != null && state.HasMethod(new MethodInstance(prop.Setter))))
                    {
                        // Handle this behavior in the next iteration
                        cascadePendingBehaviors.Add(this);
                        return visitor;
                    }
                }
                // Add this interface only for real entities, not for embedded types
                addValueHolderContainer = true;
                visitor = new EntityMetaDataHolderVisitor(visitor, metaData);
            }
            visitor = new SetCacheModificationMethodCreator(visitor);
            cascadePendingBehaviors.Add(WaitForApplyBehavior.Create(BeanContext, delegate(IClassVisitor visitor2, IBytecodeBehaviorState state2, IList<IBytecodeBehavior> remainingPendingBehaviors2,
                        IList<IBytecodeBehavior> cascadePendingBehaviors2)
                        {
                            if (addValueHolderContainer)
                            {
                                visitor2 = new InterfaceAdder(visitor2, typeof(IValueHolderContainer));
                            }
                            visitor2 = new RelationsGetterVisitor(visitor2, metaData, ValueHolderContainerHelper, PropertyInfoProvider);
                            return visitor2;
                        }));
            return visitor;
        }
    }
}